### PR TITLE
Maintenance Drone Tweaks

### PR DESCRIPTION
**Pull request by SabreML**
## What Does This PR Do

1. Allows Maintenance drones to drag construction materials such as Rods, Metal sheets and Floor tiles.
2. Shows a popup if you try using your ID on a drone.
3. Allows Maintence drones and Engiborgs to pick up Light Switches and Extinguisher Cabinet frames with their grabber.

## Why It's Good For The Game

1. Nobody likes seeing metal rods inside windows.
2. No more accidentaly instakilling a drone.
3. Self explanitory.

## Changelog
:cl:
add: Maintenance drones can now drag construction materials.
add: Warns people what will happen when they try to use their ID on a maintenance drone.
tweak: Maintenance drones can now pick up Light Switches and Extinguisher Cabinet frames.
tweak: Maintenance drones and Cyborgs can now open and close Extinguisher Cabinets.
spellcheck: Fixed light replacer grammar.
/:cl:

### CODE REVIEW

**Comment by SabreML** [comment]
image
Ah crap, hang on.

File: code/modules/mob/living/silicon/robot/drone/drone_items.dm [hunk 1]
```diff
@@ -8,7 +8,6 @@
 
 	//Has a list of items that it can hold.
 	var/list/can_hold = list(
-		/obj/item/stock_parts/cell,
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/robot/drone/drone_items.dm:11]: Why was this removed?

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 2]
```diff
@@ -154,15 +154,18 @@
 			return
 
 		else
-			user.visible_message("<span class='warning'>\the [user] swipes [user.p_their()] ID card through [src], attempting to shut it down.</span>", "<span class='warning'>You swipe your ID card through \the [src], attempting to shut it down.</span>")
+			var/confirm = null
+			confirm = alert("Using your ID on a Maintenance Drone will shut it down, are you sure you want to do this?", "Disable Drone", "Yes", "No")
+			if (confirm == ("Yes"))
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:159]: We prefer that if's are adjacent to parentheses: `if(confirm == ("Yes"))`.

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 3]
```diff
@@ -323,7 +326,12 @@
 
 /mob/living/silicon/robot/drone/start_pulling(var/atom/movable/AM)
 
-	if(istype(AM,/obj/item/pipe) || istype(AM,/obj/structure/disposalconstruct))
+	if(istype(AM,/obj/item/pipe) \
+	|| istype(AM,/obj/structure/disposalconstruct) \
+	|| istype(AM,/obj/item/stack/cable_coil) \
+	|| istype(AM,/obj/item/stack/rods) \
+	|| istype(AM,/obj/item/stack/sheet) \
+	|| istype(AM,/obj/item/stack/tile))
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:334]: I suggest making a list of all these paths, something like `pullable_drone_items`, so you can do
 `if(is_type_in_list(AM, pullable_drone_items))`.

File: code/modules/mob/living/silicon/robot/drone/drone_items.dm (hunk 1 above)
Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone_items.dm:11]: `/obj/item/stock_parts`, is also listed, so it's already there.

File: code/modules/mob/living/silicon/robot/drone/drone.dm (hunk 3 above)
Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:334]: Where should I put the list?

File: code/modules/mob/living/silicon/robot/drone/drone_items.dm (hunk 1 above)
Comment (SteelSlayer) [reply code/modules/mob/living/silicon/robot/drone/drone_items.dm:11]: Oh yeah this is fine then.

File: code/modules/mob/living/silicon/robot/drone/drone.dm (hunk 3 above)
Comment (SteelSlayer) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:334]: I was thinking as a var on drones.

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 4]
```diff
@@ -321,10 +324,20 @@
 /mob/living/silicon/robot/drone/Bumped(atom/movable/AM)
 	return
 
+var/list/pullable_drone_items = list(
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:327]: Good but you should probably declare this on drones, as right now its just hanging around. I'm not sure how this would behave actually. 

What I mean is something like this:

```js
/mob/living/silicon/robot/drone
	name = "drone"
    ... // other stuff

	var/emagged_time
    var/list/pullable_drone_items = list(
        ...
    )

Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:327]: Seems to work just fine ingame as it is. *shrug

It should be like this then?
```
        var/last_reboot
        var/emagged_time
	var/list/pullable_drone_items = list(
	/obj/item/pipe,
	/obj/structure/disposalconstruct,
	/obj/item/stack/cable_coil,
	/obj/item/stack/rods,
	/obj/item/stack/sheet,
	/obj/item/stack/tile
	)

Comment (SteelSlayer) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:327]: Yeah, but indent all the entries in the list to the right one tab. Also my indenting got messed up in my last comment, fixed it though.

**Comment by SabreML** [comment]
Woo!

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 5]
```diff
@@ -154,15 +162,18 @@
 			return
 
 		else
-			user.visible_message("<span class='warning'>\the [user] swipes [user.p_their()] ID card through [src], attempting to shut it down.</span>", "<span class='warning'>You swipe your ID card through \the [src], attempting to shut it down.</span>")
+			var/confirm = null
+			confirm = alert("Using your ID on a Maintenance Drone will shut it down, are you sure you want to do this?", "Disable Drone", "Yes", "No")
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:166]: You can meld this into one line by just doing `var/confirm = alert(blah blah)`

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 6]
```diff
@@ -154,15 +162,17 @@
 			return
 
 		else
-			user.visible_message("<span class='warning'>\the [user] swipes [user.p_their()] ID card through [src], attempting to shut it down.</span>", "<span class='warning'>You swipe your ID card through \the [src], attempting to shut it down.</span>")
+			var/confirm = alert("Using your ID on a Maintenance Drone will shut it down, are you sure you want to do this?", "Disable Drone", "Yes", "No")
```
Comment (Kyep) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:165]: Doesn't this let you swipe your ID on a drone, keep the popup on-screen, and then click 'yes' later after the drone has moved offscreen?

Also, if a check is added, doesn't this mean the drone can avoid being de-activated by moving away from you really fast?

Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:165]: It does, yes.
I thought it would be better to avoid the second situation since otherwise you could never actually deactivate one, unless you got lucky.

It also won't matter for emagged drones, since they can't be deactivated anyway.

**Comment by SabreML** [comment]
Just had another idea, how about letting drones listen in to the Engineering channel, like they can with the Binary one?

File: code/modules/mob/living/silicon/robot/drone/drone.dm [hunk 7]
```diff
@@ -154,15 +166,17 @@
...
```
Comment (AffectedArc07) [inline code/modules/mob/living/silicon/robot/drone/drone.dm:169]: This isnt going to pass until this is changed to have a range check, or a 3 seconds timeout.

Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:169]: I'll try to figure that out tomorrow, too late right now.

Comment (SabreML) [reply code/modules/mob/living/silicon/robot/drone/drone.dm:169]: Pretty sure that does it.
